File: inference/cpp/examples/generate.cpp
/**
 * 简单的文本生成程序
 *
 * 用法:
 *   ./pocket-generate model.pllm --prompt "你好"
 */

#include "pocket.h"
#include <iostream>
#include <cstring>

void print_usage(const char* program) {
    std::cout << "Usage: " << program << " <model.pllm> --prompt <text>\n\n"
              << "Options:\n"
              << "  --prompt <text>      Input prompt\n"
              << "  --max-tokens <n>     Maximum tokens to generate (default: 50)\n"
              << "  --temperature <f>    Sampling temperature (default: 0.8)\n"
              << "  --top-p <f>          Top-p sampling (default: 0.9)\n"
              << "  --top-k <n>          Top-k sampling (default: 50)\n"
              << std::endl;
}

int main(int argc, char** argv) {
    if (argc < 4) {
        print_usage(argv[0]);
        return 1;
    }

    std::string model_path = argv[1];
    std::string prompt;

    // 解析参数
    pocket::GenerateConfig config;
    config.max_new_tokens = 50;
    config.temperature = 0.8f;
    config.top_p = 0.9f;
    config.top_k = 50;

    for (int i = 2; i < argc; i++) {
        if (strcmp(argv[i], "--prompt") == 0 && i + 1 < argc) {
            prompt = argv[++i];
        } else if (strcmp(argv[i], "--max-tokens") == 0 && i + 1 < argc) {
            config.max_new_tokens = std::atoi(argv[++i]);
        } else if (strcmp(argv[i], "--temperature") == 0 && i + 1 < argc) {
            config.temperature = std::atof(argv[++i]);
        } else if (strcmp(argv[i], "--top-p") == 0 && i + 1 < argc) {
            config.top_p = std::atof(argv[++i]);
        } else if (strcmp(argv[i], "--top-k") == 0 && i + 1 < argc) {
            config.top_k = std::atoi(argv[++i]);
        }
    }

    if (prompt.empty()) {
        std::cerr << "Error: --prompt is required" << std::endl;
        return 1;
    }

    std::cout << "Loading model from " << model_path << "..." << std::endl;

    // 加载模型
    pocket::Model model;
    if (!model.load(model_path)) {
        std::cerr << "Failed to load model!" << std::endl;
        return 1;
    }

    // 加载分词器
    pocket::Tokenizer tokenizer;
    if (!tokenizer.load(model_path)) {
        std::cerr << "Failed to load tokenizer!" << std::endl;
        return 1;
    }

    // 打印模型信息
    pocket::print_model_info(model);
    std::cout << std::endl;

    // 创建生成器
    pocket::Generator generator(&model, &tokenizer);

    // 生成
    std::cout << "Prompt: " << prompt << std::endl;
    std::cout << "Generating..." << std::endl;
    std::cout << std::endl;

    std::cout << "Output: " << std::flush;

    // 设置流式输出
    config.callback = [](pocket::token_t token, void* user_data) {
        pocket::Tokenizer* tok = (pocket::Tokenizer*)user_data;
        std::cout << tok->decode(token) << std::flush;
    };
    config.callback_data = &tokenizer;

    std::string response = generator.generate(prompt, config);

    std::cout << std::endl;
    std::cout << std::endl;
    std::cout << "Done!" << std::endl;

    return 0;
}

File: inference/python/generate.py
"""
Pocket-0.1 文本生成脚本

用法:
    python inference/python/generate.py \
        --model models/pocket-0.1-pretrain \
        --prompt "从前有座山，"
"""

import argparse
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer


def main():
    parser = argparse.ArgumentParser(description='Pocket-0.1 Text Generation')
    parser.add_argument('--model', type=str, required=True, help='Path to HuggingFace model')
    parser.add_argument('--prompt', type=str, required=True, help='Input prompt')
    parser.add_argument('--device', type=str, default='cuda' if torch.cuda.is_available() else 'cpu',
                        help='Device to run on (cuda/cpu)')
    parser.add_argument('--max-new-tokens', type=int, default=100, help='Maximum new tokens to generate')
    parser.add_argument('--temperature', type=float, default=0.8, help='Sampling temperature')
    parser.add_argument('--top-p', type=float, default=0.9, help='Top-p (nucleus) sampling')
    parser.add_argument('--top-k', type=int, default=50, help='Top-k sampling')
    parser.add_argument('--num-return-sequences', type=int, default=1, help='Number of sequences to generate')
    parser.add_argument('--do-sample', action='store_true', default=True, help='Use sampling (vs greedy)')
    args = parser.parse_args()

    print("Loading model...")
    model = AutoModelForCausalLM.from_pretrained(
        args.model,
        trust_remote_code=True,
        torch_dtype=torch.float16 if args.device == 'cuda' else torch.float32,
    ).to(args.device)

    print("Loading tokenizer...")
    tokenizer = AutoTokenizer.from_pretrained(args.model, trust_remote_code=True)

    # 设置特殊 token
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token

    print(f"\n{'='*60}")
    print(f"Model: {args.model}")
    print(f"Device: {args.device}")
    print(f"Parameters: {sum(p.numel() for p in model.parameters()) / 1e6:.1f}M")
    print(f"{'='*60}\n")

    # 编码输入
    inputs = tokenizer(args.prompt, return_tensors="pt")
    inputs = {k: v.to(args.device) for k, v in inputs.items()}

    print(f"Prompt: {args.prompt}")
    print(f"\nGenerating {args.num_return_sequences} sequence(s)...\n")

    # 生成
    with torch.no_grad():
        outputs = model.generate(
            **inputs,
            max_new_tokens=args.max_new_tokens,
            temperature=args.temperature,
            top_p=args.top_p,
            top_k=args.top_k,
            do_sample=args.do_sample,
            num_return_sequences=args.num_return_sequences,
            pad_token_id=tokenizer.pad_token_id,
            eos_token_id=tokenizer.eos_token_id,
        )

    # 解码并打印
    for i, output in enumerate(outputs):
        generated_text = tokenizer.decode(output, skip_special_tokens=True)
        print(f"{'='*60}")
        print(f"Sequence {i+1}:")
        print(f"{'='*60}")
        print(generated_text)
        print()


if __name__ == "__main__":
    main()
